ExceptionInPacketCodecException(cause=java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1)

不知道发生了什么，汇报一下调用栈。 最后显示 Scheduled key refresh 不知道是不是只是网络波动导致失败了一次

```txt
[2021-07-09 14:54:25][VERB][MiraiNet]Send: wtlogin.exchange_emp(wtlogin.exchange_emp)
[2021-07-09 14:54:25][EXCE][MiraiNet]Exception in coroutine 'unnamed'
ExceptionInPacketCodecException(cause=java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler.handleExceptionInDecoding(NettyNetworkHandler.kt:63)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$ByteBufToIncomingPacketDecoder.channelRead0(NettyNetworkHandler.kt:85)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$ByteBufToIncomingPacketDecoder.channelRead0(NettyNetworkHandler.kt:75)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1
        at net.mamoe.mirai.internal.utils.crypto.TEA.doOption$decrypt(TEA.kt:228)
        at net.mamoe.mirai.internal.utils.crypto.TEA.doOption(TEA.kt:296)
        at net.mamoe.mirai.internal.utils.crypto.TEA.decrypt(TEA.kt:320)
        at net.mamoe.mirai.internal.network.components.PacketCodecImpl.parseOicqResponse(PacketCodec.kt:225)
        at net.mamoe.mirai.internal.network.components.PacketCodecImpl.decodeRaw(PacketCodec.kt:109)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$ByteBufToIncomingPacketDecoder.channelRead0(NettyNetworkHandler.kt:82)
        ... 23 more
        Suppressed: net.mamoe.mirai.internal.network.components.OicqDecodingException
        Caused by: [CIRCULAR REFERENCE: java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1]

[2021-07-09 14:54:30][VERB][MiraiNet]Send: wtlogin.exchange_emp(wtlogin.exchange_emp)
[2021-07-09 14:54:30][EXCE][MiraiNet]Exception in coroutine 'unnamed'
ExceptionInPacketCodecException(cause=java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler.handleExceptionInDecoding(NettyNetworkHandler.kt:63)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$ByteBufToIncomingPacketDecoder.channelRead0(NettyNetworkHandler.kt:85)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$ByteBufToIncomingPacketDecoder.channelRead0(NettyNetworkHandler.kt:75)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1
        at net.mamoe.mirai.internal.utils.crypto.TEA.doOption$decrypt(TEA.kt:228)
        at net.mamoe.mirai.internal.utils.crypto.TEA.doOption(TEA.kt:296)
        at net.mamoe.mirai.internal.utils.crypto.TEA.decrypt(TEA.kt:320)
        at net.mamoe.mirai.internal.network.components.PacketCodecImpl.parseOicqResponse(PacketCodec.kt:225)
        at net.mamoe.mirai.internal.network.components.PacketCodecImpl.decodeRaw(PacketCodec.kt:109)
        at net.mamoe.mirai.internal.network.impl.netty.NettyNetworkHandler$ByteBufToIncomingPacketDecoder.channelRead0(NettyNetworkHandler.kt:82)
        ... 23 more
        Suppressed: net.mamoe.mirai.internal.network.components.OicqDecodingException
        Caused by: [CIRCULAR REFERENCE: java.lang.IllegalArgumentException: data must len % 8 == 0 && len >= 16 but given -1]

[2021-07-09 14:54:35][EXCE][MiraiNet][KeyRefreshProcessor] Failed to refresh key.
kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
        at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:186)
        at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:156)
        at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:497)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
        at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:69)
        at java.base/java.lang.Thread.run(Thread.java:829)
[2021-07-09 14:54:35][INFO][MiraiNet][KeyRefreshProcessor] Scheduled key refresh in 23h 55min 0.0s.
```

--- a/mirai-core/src/commonMain/kotlin/network/components/PacketCodec.kt
+++ b/mirai-core/src/commonMain/kotlin/network/components/PacketCodec.kt
@@ -205,11 +205,12 @@ internal class PacketCodecImpl : PacketCodec {
             (client as QQAndroidClient).bot.components[EcdhInitialPublicKeyUpdater].getECDHWithPublicKey()
         return when (encryptionMethod) {
             4 -> {
+                val size = (this.remaining - 1).toInt()
                 val data =
                     TEA.decrypt(
                         this.readBytes(),
                         ecdhWithPublicKey.keyPair.maskedShareKey,
-                        length = (this.remaining - 1).toInt()
+                        length = size
                     )
 
                 val peerShareKey =
@@ -217,11 +218,12 @@ internal class PacketCodecImpl : PacketCodec {
                 TEA.decrypt(data, peerShareKey)
             }
             3 -> {
+                val size = (this.remaining - 1).toInt();
                 // session
                 TEA.decrypt(
                     this.readBytes(),
                     client.wLoginSigInfo.wtSessionTicketKey,
-                    length = (this.remaining - 1).toInt()
+                    length = size
                 )
             }
             0 -> {
@@ -235,7 +237,8 @@ internal class PacketCodecImpl : PacketCodec {
                         TEA.decrypt(byteArrayBuffer, client.randomKey, length = size)
                     }
                 } else {
-                    TEA.decrypt(this.readBytes(), client.randomKey, length = (this.remaining - 1).toInt())
+                    val size = (this.remaining - 1).toInt()
+                    TEA.decrypt(this.readBytes(), client.randomKey, length = size)
                 }
             }
             else -> error("Illegal encryption method. expected 0 or 4, got $encryptionMethod")

